cmd/pinniped-proxy/src/expired_value_cache.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright 2020-2022 the Kubeapps contributors.
+// SPDX-License-Identifier: Apache-2.0
+
+// NOTE: This may be better in the upstream cached repository. I've created
+// an issue to check with the author, otherwise it can stay here or I may
+// spin it off to a separate crate.
+// https://github.com/jaemk/cached/issues/115
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use cached::Cached;
+
+pub trait CanExpire {
+    fn is_expired(&self) -> bool;
+}
+
+// Implement our ExpiredValueCache which uses the returned values own
+// trait to determine whether it has expired, rather than a set timestamp.
+struct ExpiredValueCache<K: Hash + Eq, V: CanExpire> {
+    store: HashMap<K, V>,
+    capacity: usize,
+    hits: u64,
+    misses: u64,
+}
+impl<K: Hash + Eq, V: CanExpire> ExpiredValueCache<K, V> {
+    pub fn with_capacity(size: usize) -> ExpiredValueCache<K, V> {
+        ExpiredValueCache {
+            store: HashMap::with_capacity(size),
+            capacity: size,
+            hits: 0,
+            misses: 0,
+        }
+    }
+}
+
+impl<K: Hash + Eq, V: CanExpire> Cached<K, V> for ExpiredValueCache<K, V> {
+    fn cache_get(&mut self, k: &K) -> Option<&V> {
+        let optv = self.store.get(k);
+        // If it has expired, delete the cached entry and return none.
+        match optv {
+            Some(v) => {
+                if v.is_expired() {
+                    self.cache_remove(k);
+                    self.misses += 1;
+                    return None;
+                }
+            },
+            None => {
+                self.misses += 1;
+                return None
+            },
+        }
+        // We cannot simply return optv here because the ownership of the value
+        // moved to the self.cache_remove(k), so optv cannot be referenced after
+        // that line.
+        self.hits += 1;
+        self.store.get(k)
+    }
+
+    fn cache_get_mut(&mut self, k: &K) -> Option<&mut V> {
+        let optv = self.store.get(k);
+        // If it has expired, delete the cached entry and return none.
+        match optv {
+            Some(v) => {
+                if v.is_expired() {
+                    self.cache_remove(k);
+                    self.misses += 1;
+                    return None;
+                }
+            },
+            None => {
+                self.misses += 1;
+                return None
+            }
+        }
+        self.hits += 1;
+        self.store.get_mut(k)
```

Comment:
It seems that the only difference between this function and the `cache_get` above is this line.
Why not refactoring the common part out so that both `cache_get` and `cache_get_mut` use it?

Reply:
I'm not into Rust (for now!).
If there is no straight forward way of refactoring it, I don't think it is worth investing the effort.
Thanks!